warehouse/query-core/src/main/java/datawave/query/jexl/visitors/MergeAdjacentJunctionsVisitor.java

```diff
@@ -0,0 +1,124 @@
+package datawave.query.jexl.visitors;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.jexl3.parser.ASTAndNode;
+import org.apache.commons.jexl3.parser.ASTOrNode;
+import org.apache.commons.jexl3.parser.ASTReferenceExpression;
+import org.apache.commons.jexl3.parser.JexlNode;
+import org.apache.commons.jexl3.parser.JexlNodes;
+import org.apache.commons.jexl3.parser.ParserTreeConstants;
+
+import datawave.query.jexl.nodes.QueryPropertyMarker;
+
+/**
+ * A visitor that merges adjacent {@link ASTAndNode} and {@link ASTOrNode}
+ * <p>
+ * It is easier to run multiple visitors in sequence than it is to merge all functionality into a single visitor.
+ */
+public class MergeAdjacentJunctionsVisitor extends BaseVisitor {
+
+    public static JexlNode merge(JexlNode node) {
+        MergeAdjacentJunctionsVisitor visitor = new MergeAdjacentJunctionsVisitor();
+        node.jjtAccept(visitor, null);
+        return node;
+    }
+
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        node.childrenAccept(this, data);
+        mergeSimilarChildren(node);
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        node.childrenAccept(this, data);
+        mergeSimilarChildren(node);
+        return data;
+    }
+
+    /**
+     * Merges any children of the provided JexlNode that match the node id
+     *
+     * @param node
+     *            the JexlNode
+     */
+    protected void mergeSimilarChildren(JexlNode node) {
+
+        if (QueryPropertyMarker.findInstance(node).isAnyType()) {
+            // do not operate on any marker type
+            return;
+        }
+
+        boolean adjacent = false;
+        int target = JexlNodes.id(node);
+
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            JexlNode child = node.jjtGetChild(i);
+            if (JexlNodes.id(child) == target && !QueryPropertyMarker.findInstance(child).isAnyType()) {
+                adjacent = true;
+                break;
+            }
+        }
+
+        if (adjacent) {
+            List<JexlNode> children = new ArrayList<>();
+            for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+                JexlNode child = node.jjtGetChild(i);
+                if (JexlNodes.id(child) == target && !QueryPropertyMarker.findInstance(child).isAnyType()) {
+                    children.addAll(List.of(JexlNodes.getChildren(child)));
+                } else {
+                    children.add(child);
+                }
+            }
+            JexlNodes.setChildren(node, children.toArray(new JexlNode[0]));
+        } else if (node.jjtGetNumChildren() == 1 && node.jjtGetParent() != null) {
+            // junction with single child should be eliminated
+            JexlNodes.swap(node.jjtGetParent(), node, node.jjtGetChild(0));
+        }
+    }
+
+    @Override
+    public Object visit(ASTReferenceExpression node, Object data) {
```

Comment:
Would like to see a comment here detailing how this merging similar junctions between parents and wrapped junction grandchildren. If that's not what's happening here, then please correct me.